codegen/src/types.rs

```diff
@@ -1356,3 +1410,64 @@ fn break_cycles(messages: &mut [Message], leaf_messages: &mut Vec<String>) {
 fn tag(number: u32, typ: &FieldType, packed: bool) -> u32 {
     number << 3 | typ.wire_type_num(packed)
 }
+
+/// "" is ("",""), "a" is ("","a"), "a.b" is ("a"."b"), and so forth.
+fn split_package(package: &str) -> (&str,&str) {
+    if package.is_empty() {
+        ("","")
+    } else {
+        if let Some(i) = package.rfind('.') {
+            (&package[0..i],&package[i+1..])
+        } else {
+            ("",package)
+        }
+    }
+}
+
+/// Given a file path, create or update the mod.rs file within its folder
+fn update_mod_file(path: &Path) -> Result<()> {
+    let mut file = path.to_path_buf();
+    use std::fs::OpenOptions;
+    use std::io::prelude::*;
+    let name = file.file_stem().unwrap().to_string_lossy().to_string();
+    file.pop();
+    file.push("mod.rs");
+    let matches = "pub mod ";
+    let mut present = false;
+    let mut exists = false;
+    if let Ok(f) = File::open(&file) {
+        exists = true;
+        for line in BufReader::new(f).lines() {
+            let line = line?;
+            if let Some(i) = line.find(matches) {
+                let rest = &line[i+matches.len()..line.len()-1];
+                if rest == name {
```

Comment:
The slice for `pub use foobar;` is going to be "foobar", which is matched exactly. This feels less than elegant, but it works as intended.  (that `line.len()-1` feels very hacky)